drivers/crypto/crypto_nrf52_ecb.c

```diff
@@ -0,0 +1,169 @@
+	u8_t *ciphertext;     ///< Ciphertext: Starts at ecb_data + 32 bytes.
+	int in_use;
+};
+
+static struct nrf52_ecb_drv_state_t nrf52_ecb_drv_state;
+
+static int do_ecb_encrypt(struct cipher_ctx *ctx, struct cipher_pkt *pkt);
+
+static int nrf52_ecb_driver_init(struct device *dev)
+{
+	struct nrf52_ecb_drv_state_t *state = &nrf52_ecb_drv_state;
+
+	ARG_UNUSED(dev);
+
+	//use array index for offsets?
+	state->key = state->ecb_data + ECBDATA_KEY_OFFSET;
+	state->cleartext = state->ecb_data + ECBDATA_CLEARTEXT_OFFSET;
+	state->ciphertext = state->ecb_data + ECBDATA_CIPHERTEXT_OFFSET;
+
+	nrf_ecb_data_pointer_set(ECB_PERIPHERAL, state->ecb_data);
+	return 0;
+}
+
+static int nrf52_ecb_query_caps(struct device *dev)
+{
+	return (CAP_RAW_KEY | CAP_SEPARATE_IO_BUFS | CAP_SYNC_OPS);
+}
+
+static int nrf52_ecb_session_setup(struct device *dev, struct cipher_ctx *ctx,
+		     enum cipher_algo algo, enum cipher_mode mode,
+		     enum cipher_op op_type)
+{
+	ARG_UNUSED(dev);
+
+	if (algo != CRYPTO_CIPHER_ALGO_AES) {
+		LOG_ERR("unsupported algo");
+		return -EINVAL;
+	}
+
+	if (!(ctx->flags & CAP_SYNC_OPS)) {
+		LOG_ERR("Async not supported by this driver");
+		return -EINVAL;
+	}
+
+	if (ctx->keylen != AES_KEY_SIZE) {
+		/* driver supports 128 bits */
+		LOG_ERR("unsupported key size");
+		return -EINVAL;
+	}
+
+	if (ctx->key.bit_stream == NULL) {
+		LOG_ERR("no key provided");
+		return -EINVAL;
+	}
+
+	if (op_type == CRYPTO_CIPHER_OP_ENCRYPT) {
+		switch (mode) {
+		case CRYPTO_CIPHER_MODE_ECB:
+			ctx->ops.block_crypt_hndlr = do_ecb_encrypt;
+			break;
+		default:
+			LOG_ERR("unsupported mode");
+			return -EINVAL;
+		}
+	} else {
+		LOG_ERR("unsupported operation");
+		return -EINVAL;
+	}
+
+	ctx->ops.cipher_mode = mode;
+
+	if (nrf52_ecb_drv_state.in_use) {
+		LOG_WRN("Peripheral in use");
+		return -EBUSY;
+	} else {
+		nrf52_ecb_drv_state.in_use = 1;
+	}
+
+	memcpy(nrf52_ecb_drv_state.key, ctx->key.bit_stream, AES_KEY_SIZE);
+	ctx->drv_sessn_state = &nrf52_ecb_drv_state;
+
+	return 0;
+}
+
+static int nrf52_ecb_session_free(struct device *dev, struct cipher_ctx *sessn)
+{
+	struct nrf52_ecb_drv_state_t *state;
+
+	ARG_UNUSED(dev);
+	state = (struct nrf52_ecb_drv_state_t*)sessn->drv_sessn_state;
+	state->in_use = 0;
+
+	return 0;
+}
+
+int do_ecb_encrypt(struct cipher_ctx *ctx, struct cipher_pkt *pkt)
+{
+	struct nrf52_ecb_drv_state_t *state;
+	u32_t counter = 0x1000000;
+
+	state = (struct nrf52_ecb_drv_state_t*)ctx->drv_sessn_state;
+	if (pkt->in_len != AES_BLOCK_SIZE) {
+		LOG_ERR("unsupported input block size");
+		return -EINVAL;
+	}
+	if (pkt->out_buf_max < pkt->in_len) {
+		LOG_ERR("output buff too small");
+		return -EINVAL;
+	}
+
+	if (pkt->in_buf != state->cleartext) {
+		memcpy(state->cleartext, pkt->in_buf, AES_BLOCK_SIZE);
+	}
+
+	nrf_ecb_event_clear(ECB_PERIPHERAL, NRF_ECB_EVENT_ENDECB);
+	nrf_ecb_task_trigger(ECB_PERIPHERAL, NRF_ECB_TASK_STARTECB);
+	while (nrf_ecb_event_check(ECB_PERIPHERAL, NRF_ECB_EVENT_ENDECB) == false) {
```

Comment:
Why not use the ECB interrupt? You could then provide both asynchronous and synchronous operation. The latter with waiting on a semaphore here instead of this loop with artificial timeout checking with a counter.

Reply:
Yes, I agree that using ECB interrupt is a more elegant solution and timeout checking with a counter is artificial. On the other hand the task that this loop is waiting for takes just 7.2 us. I don't think we can benefit from returning control to OS for such short time.
My idea is to remove counter checking but to test all possible results of the operation (encryption _end_ or _error_), not just the positive one like it is now (encryption _end_) so that there is no threat that it stays in the loop for longer.